contracts/admin/src/common/mod.rs

```diff
@@ -0,0 +1,124 @@
+use std::error::Error;
+
+use serde::{Deserialize, Serialize};
+
+use platform::batch::Batch;
+use sdk::{
+    cosmwasm_std::{Addr, Binary, QuerierWrapper, WasmMsg},
+    schemars::{self, JsonSchema},
+};
+
+use crate::{
+    common::type_defs::MaybeMigrateContract,
+    error::{ContractError, Never},
+};
+
+pub(crate) mod type_defs;
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, JsonSchema)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+pub struct CodeIdWithMigrateMsg<M> {
+    pub code_id: u64,
+    pub migrate_msg: M,
+}
+
+pub(crate) trait ValidateAddresses {
+    fn validate(&self, querier: &QuerierWrapper<'_>) -> Result<(), ContractError>;
+}
+
+pub(crate) trait MigrateContracts {
+    type GatSelf<T>;
```

Comment:
what is `Gat` like?

Reply:
Genetic Associated Type, in short GAT, which then converted to PascalCase gives us this.

Reply:
you meant Generic not Genetic

Reply:
Checked on stabilization report and it was stabilized in Rust 1.65. Going to have to resort back to using a traditional associated type.